src/LinearAlgebra/SparseMatrix/CsrMatrix.hpp

```diff
@@ -0,0 +1,172 @@
+      Sparse(index_type n,
+             index_type m,
+             index_type nnz,
+             bool       symmetric,
+             bool       expanded);
+      virtual ~Sparse();
+
+      // accessors
+      index_type   getNumRows();
+      index_type   getNumColumns();
+      index_type   getNnz();
+      SparseFormat getSparseFormat() const;
+
+      bool symmetric();
+      bool expanded();
+      void setSymmetric(bool symmetric);
+      void setExpanded(bool expanded);
+      void setNnz(index_type nnz_new); // for resetting when removing duplicates
+      int  setUpdated(memory::MemorySpace what);
+
+      virtual index_type* getRowData(memory::MemorySpace memspace) = 0;
+      virtual index_type* getColData(memory::MemorySpace memspace) = 0;
+      virtual real_type*  getValues(memory::MemorySpace memspace)  = 0;
+
+      virtual int copyDataFrom(const index_type*   row_data,
+                               const index_type*   col_data,
+                               const real_type*    val_data,
+                               memory::MemorySpace memspaceIn,
+                               memory::MemorySpace memspaceOut) = 0;
+      virtual int copyDataFrom(const index_type*   row_data,
+                               const index_type*   col_data,
+                               const real_type*    val_data,
+                               index_type          new_nnz,
+                               memory::MemorySpace memspaceIn,
+                               memory::MemorySpace memspaceOut) = 0;
+
+      virtual int allocateMatrixData(memory::MemorySpace memspace) = 0;
+      int         setDataPointers(index_type*         row_data,
+                                  index_type*         col_data,
+                                  real_type*          val_data,
+                                  memory::MemorySpace memspace);
+
+      int destroyMatrixData(memory::MemorySpace memspace);
+
+      virtual void print(std::ostream& file_out, index_type indexing_base) = 0;
+
+      virtual int syncData(memory::MemorySpace memspaceOut) = 0;
+
+      // update Values just updates values; it allocates if necessary.
+      // values have the same dimensions between different formats
+      virtual int copyValues(const real_type*    new_vals,
+                             memory::MemorySpace memspaceIn,
+                             memory::MemorySpace memspaceOut);
+
+      // set new values just sets the pointer, use caution.
+      virtual int setValuesPointer(real_type*          new_vals,
+                                   memory::MemorySpace memspace);
+
+    protected:
+      SparseFormat sparse_format_{NONE}; ///< Matrix format
+      index_type   n_{0};                ///< number of rows
+      index_type   m_{0};                ///< number of columns
+      index_type   nnz_{0};              ///< number of non-zeros
+
+      bool is_symmetric_{false}; ///< symmetry flag
+      bool is_expanded_{false};  ///< "expanded" flag
```

Comment:
Remove as GridKit is dealing with general matrices and there is no need for matrices stored in a symmetric form.